### PR TITLE
bpo-30547: Fix multiple reference leaks

### PR DESCRIPTION
**Pull request by matrixise**
Fix regressions introduced by:

- bpo-18520: commit 8fea252a507024edf00d5d98881d22dc8799a8d3
- bpo-22257: commits 1abcf6700b4da6207fe859de40c6c1bada6b4fec and 6b4be195cd8868b76eb6fbe166acc39beee8ce36

Fixed with @haypo and @mlouielu

### CODE REVIEW

**Review by vstinner — COMMENTED** [review]
In the commit message, can you please use the syntax:

```
Co-Authored-By: full name <name@example.com>
```

Rather than GitHub names? For me, it's:

```
Co-Authored-By: Victor Stinner <victor.stinner@gmail.com>
```

**Review by vstinner — COMMENTED** [review]
Commit message:
"Introduced by 1abcf67"

Hum, it was also introduced by 6b4be195cd8868b76eb6fbe166acc39beee8ce36. I propose:

```
Fix regressions introduced by:

- bpo-18520: commit 8fea252a507024edf00d5d98881d22dc8799a8d3
- bpo-22257: commits 1abcf6700b4da6207fe859de40c6c1bada6b4fec and 6b4be195cd8868b76eb6fbe166acc39beee8ce36
```

**Review by vstinner — APPROVED** [review]
LGTM, let's see if all tests pass ;-)

File: Python/sysmodule.c [hunk 1]
```diff
@@ -2139,7 +2139,7 @@ _PySys_EndInit(PyObject *sysdict)
     else {
         Py_INCREF(warnoptions);
     }
-    SET_SYS_FROM_STRING_BORROW_INT_RESULT("warnoptions", warnoptions);
+    SET_SYS_FROM_STRING_INT_RESULT("warnoptions", warnoptions);
```
Comment (vstinner) [inline Python/sysmodule.c]: test_capi does crash with this change :-(

**Review by vstinner — CHANGES_REQUESTED** [review]
I suggest you to revert the warnoptions change and let it open, since it seems tricky to fix it.

File: Python/sysmodule.c (hunk 1 above)
Comment (matrixise) [reply Python/sysmodule.c]: ok thanks

Comment (matrixise) [reply Python/sysmodule.c]: reverted

**Review by vstinner — CHANGES_REQUESTED** [review]
You reverted the warnoptions change, but you forgot to update the commit message: your change is not related to "- bpo-18520: commit 8fea252".

**Review by vstinner — APPROVED** [review]
Ok, it should now pass on Travis CI but it doesn't fix the refleak in warnoptions. It's a compromise :-)

**Pull request merged**